crates/vm/levm/src/gas_cost.rs

```diff
@@ -748,6 +760,103 @@ pub fn fake_exponential(factor: u64, numerator: u64, denominator: u64) -> Result
 
+pub fn sha2_256(data_size: u64) -> Result<U256, OutOfGasError> {
+    precompile(data_size, SHA2_256_STATIC_COST, SHA2_256_DYNAMIC_BASE)
+}
+
+pub fn ripemd_160(data_size: u64) -> Result<U256, OutOfGasError> {
+    precompile(data_size, RIPEMD_160_STATIC_COST, RIPEMD_160_DYNAMIC_BASE)
+}
+
+pub fn identity(data_size: u64) -> Result<U256, OutOfGasError> {
+    precompile(data_size, IDENTITY_STATIC_COST, IDENTITY_DYNAMIC_BASE)
+}
+
+pub const MODEXP_STATIC_COST: u64 = 0;
+pub const MODEXP_DYNAMIC_BASE: u64 = 200;
+pub const MODEXP_DYNAMIC_QUOTIENT: u64 = 3;
+
+pub fn modexp(
+    exponent: U256,
+    base_size: u64,
+    exponent_size: u64,
+    modulus_size: u64,
+) -> Result<u64, VMError> {
+    let max_length = base_size.max(modulus_size);
+    let words = (max_length
+        .checked_add(7)
+        .ok_or(OutOfGasError::GasCostOverflow)?)
+        / WORD_SIZE_IN_BYTES_U64;
+    let multiplication_complexity = words.checked_pow(2).ok_or(OutOfGasError::GasCostOverflow)?;
+
+    let mut iteration_count: u64 = 0;
+    if exponent_size <= WORD_SIZE_IN_BYTES_U64 && exponent.is_zero() {
+        iteration_count = 0;
+    } else if exponent_size <= WORD_SIZE_IN_BYTES_U64 {
+        iteration_count = exponent
+            .bits()
+            .checked_sub(1)
+            .ok_or(InternalError::ArithmeticOperationUnderflow)?
+            .try_into()
+            .map_err(|_| InternalError::ConversionError)?;
+    } else if exponent_size > WORD_SIZE_IN_BYTES_U64 {
+        iteration_count = 8u64
+            .checked_mul(
+                exponent_size
+                    .checked_sub(WORD_SIZE_IN_BYTES_U64)
+                    .ok_or(InternalError::ArithmeticOperationUnderflow)?,
+            )
+            .ok_or(InternalError::ArithmeticOperationOverflow)?
+            .checked_add(
+                (exponent
+                    & (2usize
+                        .checked_pow(256)
+                        .ok_or(InternalError::ArithmeticOperationOverflow)?)
+                    .checked_sub(1)
+                    .ok_or(InternalError::ArithmeticOperationOverflow)?
+                    .into())
+                .bits()
+                .checked_sub(1)
+                .ok_or(InternalError::ArithmeticOperationUnderflow)?
+                .try_into()
+                .map_err(|_| InternalError::ConversionError)?,
+            )
+            .ok_or(InternalError::ArithmeticOperationOverflow)?;
+    }
+
+    let calculate_iteration_count = iteration_count.max(1);
+
+    let static_gas = MODEXP_STATIC_COST;
+
+    let dynamic_gas = MODEXP_DYNAMIC_BASE.max(
+        multiplication_complexity
+            .checked_mul(calculate_iteration_count)
+            .ok_or(OutOfGasError::GasCostOverflow)?
+            / MODEXP_DYNAMIC_QUOTIENT,
+    );
+
+    Ok(static_gas
+        .checked_add(dynamic_gas)
+        .ok_or(OutOfGasError::GasCostOverflow)?)
+}
+
+fn precompile(data_size: u64, static_cost: u64, dynamic_base: u64) -> Result<U256, OutOfGasError> {
```

Comment:
If the rest of the precompiles' gas cost can fit in this function then it is a great idea to have it